### DISCUSSION TITLE
What is an environment in Janet for?

### DISCUSSION THREAD

**Discussion by uvtc** [janet-lang/janet#748] [General]
What is the role played by an *environment* (the table returned by `(curenv)`)? What are they for in Janet?

In:

~~~
$ janet
Janet 1.16.1-local linux/x64 - '(doc)' for help

repl:1:> (curenv)
@{_ @{:value <cycle 0>}}

repl:2:> (def a 1)
1

repl:3:> (curenv)
@{_ @{:value <cycle 0>} a @{:source-map ("repl" 2 1) :value 1}}
~~~

what does `<cycle 0>` mean?

**Comment by sogaiu** [comment]
Possibly among other things, I think an environment is used for resolution of some names, e.g.:
```
repl:1:> (curenv)
@{_ @{:value <cycle 0>}}
repl:2:> a
repl:2:1: compile error: unknown symbol a
repl:3:> (def a 1)
1
repl:4:> (curenv)
@{_ @{:value <cycle 0>} a @{:source-map ("repl" 3 1) :value 1}}
repl:5:> a
1
```
As for `<cycle 0>`, I think that's the pretty printer's way of telling you there is a "cyclic" thing going on: https://github.com/janet-lang/janet/blob/61769c8f165762b85454757b2ce5a5ffb555c21c/src/core/pp.c#L492-L497

**Reply by uvtc** [reply]
Hm. Thanks. Re `<cycle 0>`, I see that something cyclic indeed happens when I try to look at the value in the curenv table for key `'_`:

~~~
$ janet
Janet 1.16.1-local linux/x64 - '(doc)' for help

repl:1:> (get (curenv) '_)
nil

repl:2:> (get (curenv) '_)
@{}

repl:3:> (get (curenv) '_)
@{:value @{}}

repl:4:> (get (curenv) '_)
@{:value @{:value @{}}}

repl:5:> (get (curenv) '_)
@{:value @{:value @{:value @{}}}}
~~~

**Comment by subsetpark** [comment]
The current execution environment is realized by maintaining a series of simple tables, each of which links back to the previous environment in the stack as a table prototype. So when you evaluate any symbol, you’re essentially doing a `get` for it in the current environment; on the other hand, if you print out the current environment you might be slightly confused because you’ll only see what’s in the current frame (that’s just how tables work in general).

`_` is a symbol that’s automatically maintained by the repl, specifically (it’s not defined when executing a script or running a program). It’s bound to the value of the most recent expression entered in the repl. So if you evaluate `(+ 2 2)` you’ll get back 4, and `_` will also be bound to 4 so you can conveniently access that value in the next REPL line without having to retype the whole expression.

therefore, if you evaluate `(curenv)` you will both:

- associate _ with the current environment table, inside that same table;
- print out that table.

The pretty printer avoids an infinite loop on displaying the table inside of itself with the cycle notation.

**Comment by subsetpark** [comment]
To demonstrate what I mean, here's an example of printing out the curenv *without* having bound `_` to itself:

```
repl:3:> :foo
:foo
repl:4:> (do
repl:5:(> (pp (curenv))
repl:6:(> :bar)
@{_ @{:value :foo}}
:bar
repl:7:> 
```

**Comment by subsetpark** [comment]
And to illustrate the (arguably more relevant) question of what environments *are* at all, you can evaluate `(table/getproto (curenv))`; from the REPL; you'll see the environment created by `boot.janet`.

**Comment by uvtc** [comment]
Thanks, @subsetpark ! It sounds like the primary purpose of the environment table is for looking up bindings.

I realize that Janet of course does this automatically for us, but I see that it also provides some functions for inspecting and altering environments if we need to.

It reminds me a little of environment variables in the shell (bash), since they (shell env vars) let us define names for the current shell.
